File: auth-service/src/trading/models/GridState.js
import db from '../db.js';
import { v4 as uuidv4 } from 'uuid';

/**
 * GridState - stan algorytmu GRID dla konkretnego zlecenia
 */
export class GridState {
  constructor(data = {}) {
    this.id = data.id || uuidv4();
    this.walletAddress = data.walletAddress || data.wallet_address;
    this.orderId = data.orderId || data.order_id;
    this.currentFocusPrice = data.currentFocusPrice ?? data.current_focus_price ?? 0;
    this.focusLastUpdated = data.focusLastUpdated || data.focus_last_updated;
    this.buyTrendCounter = data.buyTrendCounter ?? data.buy_trend_counter ?? 0;
    this.sellTrendCounter = data.sellTrendCounter ?? data.sell_trend_counter ?? 0;
    this.nextBuyTarget = data.nextBuyTarget ?? data.next_buy_target;
    this.nextSellTarget = data.nextSellTarget ?? data.next_sell_target;
    this.openPositionIds = this._parseJson(data.openPositionIds || data.open_position_ids || '[]');
    this.openSellPositionIds = this._parseJson(data.openSellPositionIds || data.open_sell_position_ids || '[]');
    this.totalProfit = data.totalProfit ?? data.total_profit ?? 0;
    this.totalBuyTransactions = data.totalBuyTransactions ?? data.total_buy_transactions ?? 0;
    this.totalSellTransactions = data.totalSellTransactions ?? data.total_sell_transactions ?? 0;
    this.totalBoughtValue = data.totalBoughtValue ?? data.total_bought_value ?? 0;
    this.totalSoldValue = data.totalSoldValue ?? data.total_sold_value ?? 0;
    this.isActive = data.isActive ?? data.is_active ?? true;
    this.lastUpdated = data.lastUpdated || data.last_updated;
    this.createdAt = data.createdAt || data.created_at;
    this.lastKnownPrice = data.lastKnownPrice ?? data.last_known_price;
    this.lastPriceUpdate = data.lastPriceUpdate || data.last_price_update;
  }

  _parseJson(value) {
    if (Array.isArray(value)) return value;
    try {
      return JSON.parse(value || '[]');
    } catch {
      return [];
    }
  }

  save() {
    const now = new Date().toISOString();
    this.lastUpdated = now;
    if (!this.createdAt) this.createdAt = now;

    const stmt = db.prepare(`
      INSERT OR REPLACE INTO grid_states (
        id, wallet_address, order_id, current_focus_price, focus_last_updated,
        buy_trend_counter, sell_trend_counter, next_buy_target, next_sell_target,
        open_position_ids, open_sell_position_ids, total_profit,
        total_buy_transactions, total_sell_transactions, total_bought_value,
        total_sold_value, is_active, last_updated, created_at,
        last_known_price, last_price_update
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);

    stmt.run(
      this.id,
      this.walletAddress,
      this.orderId,
      this.currentFocusPrice,
      this.focusLastUpdated,
      this.buyTrendCounter,
      this.sellTrendCounter,
      this.nextBuyTarget,
      this.nextSellTarget,
      JSON.stringify(this.openPositionIds),
      JSON.stringify(this.openSellPositionIds),
      this.totalProfit,
      this.totalBuyTransactions,
      this.totalSellTransactions,
      this.totalBoughtValue,
      this.totalSoldValue,
      this.isActive ? 1 : 0,
      this.lastUpdated,
      this.createdAt,
      this.lastKnownPrice,
      this.lastPriceUpdate
    );

    return this;
  }

  static findByWalletAndOrderId(walletAddress, orderId) {
    const row = db.prepare(
      'SELECT * FROM grid_states WHERE wallet_address = ? AND order_id = ?'
    ).get(walletAddress, orderId);
    
    return row ? new GridState(row) : null;
  }

  static findAllActive() {
    const rows = db.prepare('SELECT * FROM grid_states WHERE is_active = 1').all();
    return rows.map(row => new GridState(row));
  }

  static findAllByWallet(walletAddress) {
    const rows = db.prepare('SELECT * FROM grid_states WHERE wallet_address = ?').all(walletAddress);
    return rows.map(row => new GridState(row));
  }

  toJSON() {
    return {
      id: this.id,
      walletAddress: this.walletAddress,
      orderId: this.orderId,
      currentFocusPrice: this.currentFocusPrice,
      focusLastUpdated: this.focusLastUpdated,
      buyTrendCounter: this.buyTrendCounter,
      sellTrendCounter: this.sellTrendCounter,
      nextBuyTarget: this.nextBuyTarget,
      nextSellTarget: this.nextSellTarget,
      openPositionIds: this.openPositionIds,
      openSellPositionIds: this.openSellPositionIds,
      totalProfit: this.totalProfit,
      totalBuyTransactions: this.totalBuyTransactions,
      totalSellTransactions: this.totalSellTransactions,
      totalBoughtValue: this.totalBoughtValue,
      totalSoldValue: this.totalSoldValue,
      isActive: this.isActive,
      lastUpdated: this.lastUpdated,
      createdAt: this.createdAt,
      lastKnownPrice: this.lastKnownPrice,
      lastPriceUpdate: this.lastPriceUpdate
    };
  }
}

export default GridState;

File: GridBot/trading-engine/src/main/java/com/gridbot/model/GridState.java
package com.gridbot.model;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.Instant;

/**
 * Stan algorytmu GRID dla konkretnego zlecenia
 */
@Entity
@Table(name = "grid_states")
public class GridState {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;
    private String walletAddress;
    private String orderId;
    private BigDecimal currentFocusPrice;
    private Instant focusLastUpdated;
    private int buyTrendCounter;
    private int sellTrendCounter;
    private BigDecimal nextBuyTarget;
    private BigDecimal nextSellTarget;
    
    @Column(length = 10000)
    private String openPositionIds; // JSON array stored as string
    
    @Column(length = 10000)
    private String openSellPositionIds; // JSON array stored as string
    
    private BigDecimal totalProfit;
    private int totalBuyTransactions;
    private int totalSellTransactions;
    private BigDecimal totalBoughtValue;
    private BigDecimal totalSoldValue;
    private boolean isActive;
    private Instant lastUpdated;
    private Instant createdAt;
    private BigDecimal lastKnownPrice;
    private Instant lastPriceUpdate;

    public GridState() {
        this.openPositionIds = "[]";
        this.openSellPositionIds = "[]";
        this.totalProfit = BigDecimal.ZERO;
        this.totalBoughtValue = BigDecimal.ZERO;
        this.totalSoldValue = BigDecimal.ZERO;
    }

    // Getters
    public String getId() { return id; }
    public String getWalletAddress() { return walletAddress; }
    public String getOrderId() { return orderId; }
    public BigDecimal getCurrentFocusPrice() { return currentFocusPrice; }
    public Instant getFocusLastUpdated() { return focusLastUpdated; }
    public int getBuyTrendCounter() { return buyTrendCounter; }
    public int getSellTrendCounter() { return sellTrendCounter; }
    public BigDecimal getNextBuyTarget() { return nextBuyTarget; }
    public BigDecimal getNextSellTarget() { return nextSellTarget; }
    public String getOpenPositionIds() { return openPositionIds; }
    public String getOpenSellPositionIds() { return openSellPositionIds; }
    public BigDecimal getTotalProfit() { return totalProfit; }
    public int getTotalBuyTransactions() { return totalBuyTransactions; }
    public int getTotalSellTransactions() { return totalSellTransactions; }
    public BigDecimal getTotalBoughtValue() { return totalBoughtValue; }
    public BigDecimal getTotalSoldValue() { return totalSoldValue; }
    public boolean isActive() { return isActive; }
    public Instant getLastUpdated() { return lastUpdated; }
    public Instant getCreatedAt() { return createdAt; }
    public BigDecimal getLastKnownPrice() { return lastKnownPrice; }
    public Instant getLastPriceUpdate() { return lastPriceUpdate; }

    // Setters
    public void setId(String id) { this.id = id; }
    public void setWalletAddress(String walletAddress) { this.walletAddress = walletAddress; }
    public void setOrderId(String orderId) { this.orderId = orderId; }
    public void setCurrentFocusPrice(BigDecimal currentFocusPrice) { this.currentFocusPrice = currentFocusPrice; }
    public void setFocusLastUpdated(Instant focusLastUpdated) { this.focusLastUpdated = focusLastUpdated; }
    public void setBuyTrendCounter(int buyTrendCounter) { this.buyTrendCounter = buyTrendCounter; }
    public void setSellTrendCounter(int sellTrendCounter) { this.sellTrendCounter = sellTrendCounter; }
    public void setNextBuyTarget(BigDecimal nextBuyTarget) { this.nextBuyTarget = nextBuyTarget; }
    public void setNextSellTarget(BigDecimal nextSellTarget) { this.nextSellTarget = nextSellTarget; }
    public void setOpenPositionIds(String openPositionIds) { this.openPositionIds = openPositionIds; }
    public void setOpenSellPositionIds(String openSellPositionIds) { this.openSellPositionIds = openSellPositionIds; }
    public void setTotalProfit(BigDecimal totalProfit) { this.totalProfit = totalProfit; }
    public void setTotalBuyTransactions(int totalBuyTransactions) { this.totalBuyTransactions = totalBuyTransactions; }
    public void setTotalSellTransactions(int totalSellTransactions) { this.totalSellTransactions = totalSellTransactions; }
    public void setTotalBoughtValue(BigDecimal totalBoughtValue) { this.totalBoughtValue = totalBoughtValue; }
    public void setTotalSoldValue(BigDecimal totalSoldValue) { this.totalSoldValue = totalSoldValue; }
    public void setActive(boolean active) { isActive = active; }
    public void setLastUpdated(Instant lastUpdated) { this.lastUpdated = lastUpdated; }
    public void setCreatedAt(Instant createdAt) { this.createdAt = createdAt; }
    public void setLastKnownPrice(BigDecimal lastKnownPrice) { this.lastKnownPrice = lastKnownPrice; }
    public void setLastPriceUpdate(Instant lastPriceUpdate) { this.lastPriceUpdate = lastPriceUpdate; }

    // Builder
    public static Builder builder() { return new Builder(); }

    public static class Builder {
        private final GridState s = new GridState();
        public Builder id(String v) { s.id = v; return this; }
        public Builder walletAddress(String v) { s.walletAddress = v; return this; }
        public Builder orderId(String v) { s.orderId = v; return this; }
        public Builder currentFocusPrice(BigDecimal v) { s.currentFocusPrice = v; return this; }
        public Builder focusLastUpdated(Instant v) { s.focusLastUpdated = v; return this; }
        public Builder buyTrendCounter(int v) { s.buyTrendCounter = v; return this; }
        public Builder sellTrendCounter(int v) { s.sellTrendCounter = v; return this; }
        public Builder nextBuyTarget(BigDecimal v) { s.nextBuyTarget = v; return this; }
        public Builder nextSellTarget(BigDecimal v) { s.nextSellTarget = v; return this; }
        public Builder openPositionIds(String v) { s.openPositionIds = v != null ? v : "[]"; return this; }
        public Builder openSellPositionIds(String v) { s.openSellPositionIds = v != null ? v : "[]"; return this; }
        public Builder totalProfit(BigDecimal v) { s.totalProfit = v != null ? v : BigDecimal.ZERO; return this; }
        public Builder totalBuyTransactions(int v) { s.totalBuyTransactions = v; return this; }
        public Builder totalSellTransactions(int v) { s.totalSellTransactions = v; return this; }
        public Builder totalBoughtValue(BigDecimal v) { s.totalBoughtValue = v != null ? v : BigDecimal.ZERO; return this; }
        public Builder totalSoldValue(BigDecimal v) { s.totalSoldValue = v != null ? v : BigDecimal.ZERO; return this; }
        public Builder isActive(boolean v) { s.isActive = v; return this; }
        public Builder lastUpdated(Instant v) { s.lastUpdated = v; return this; }
        public Builder createdAt(Instant v) { s.createdAt = v; return this; }
        public Builder lastKnownPrice(BigDecimal v) { s.lastKnownPrice = v; return this; }
        public Builder lastPriceUpdate(Instant v) { s.lastPriceUpdate = v; return this; }
        public GridState build() { return s; }
    }
}
